Try canonical block plugins using the Time to Read block

creativecoder opened the pull request:
## What?

Experiment to try creating a canonical block plugin, using the existing Time to Read block.

Part of https://github.com/WordPress/gutenberg/issues/58773

## Why?

**Canonical Block Plugin?**

From the https://github.com/WordPress/gutenberg/issues/58773 description:

> There's a growing subset of blocks that we may contemplate creating that are either more niche or—for various reasons—not necessarily an immediate fit for the bundled library in core...
> 
> I propose we look at a mechanism where such blocks can be designed, developed, published, and maintained by core contributors in this repository (benefits from collocation and testing infrastructure) but for them to be published as standalone blocks in the directory instead of bundled in the default library. These blocks would show up as authored by [WordPress.org]. It'd allow themes and patterns in the directory to leverage them with certainty.

**Time to Read Block**

The block already exists in Gutenberg, but has been punted from WP releases twice, over concerns it's not a good fit for the default block library (https://github.com/WordPress/gutenberg/issues/53776). But it still seems to be a popular block, so would be useful to have available in a stand-alone block plugin for easy installation from the editor (for those who don't run the Gutenberg plugin on their site).

## How?

**Time to Read block**

- Copies the existing time-to-read block into it's own directory.
- Adds readme, screenshots, etc so that the block can be released as a standalone plugin and added to the block directory.
- This PR doesn't modify the existing behavior of the block in any way, only adding what's needed to make it work as a standalone plugin, based on `create-block-plugin` scaffolding.

**Canonical plugins infrastructure**

- Adds configuration for linting, automated tests, wp-env, etc to use the new `plugins/` folder in the repo.
- Updates npm scripts to build and lint plugins.
- Uses the npm workspaces feature to enable running individual plugin scripts from the repo root (e.g. `npm run build -w plugins/time-to-read-block`).

See the `plugins/README.md` file in this PR for more details and documentation.

## Testing Instructions

- Disable the Gutenberg plugin in your development environment, to make sure the block plugin works independently.
- Build the project: `npm run build`
- If not using the included wp-env development environment, symlink the plugin folder to `wp-content/plugins/time-to-read-block` in your development environment
- Activate the plugin
- Search for the "Time to Read" block in the editor and insert it into a post or page
- Save and check the front-end of the site

## Future work

This PR does not include scripts and workflows for the following:

- Generating new plugin releases and bumping versions
- Deploying plugin updates to the plugins repo
- Automatically updating `Requires at least:` and `Requires PHP:` plugin headers when new versions of WordPress are released

Commit:
add time to read block

creativecoder committed:
Update settings to register and render block

creativecoder committed:
Updates time to read block plugin

- Rename and move plugin as time-to-read-block
- Update readme.txt and plugin headers with appropriate values
- Add screenshots
- Add phpcs and eslint configuration overrides
- Update PHP function names to use the gutenberg_ namespace
- Resolve eslint error by installing the clsx package

creativecoder committed:
Remove example-static plugin

creativecoder committed:
Adds small fixes/edits

gziolo commented:
This is a great start. Do you want it to be more tightly integrated with the monorepo? I could offer a few hints on moving some of the configurations to the root of the projects. In the long run, we could also explore improving linting overrides related to textdinaub so plugin authors can benefit from it and avoid the need to maintain their own customizations.

creativecoder commented:
> This is a great start. Do you want it to be more tightly integrated with the monorepo? I could offer a few hints on moving some of the configurations to the root of the projects.

@gziolo , thanks for taking a look! Yes, that would be great--any hints you have would be appreciated. I'd like to use this PR to experiment with monorepo integration, CI configuration, and the like. We can always break it out into more granular changes before merging, if needed.

gziolo commented on plugins/time-to-read-block/.eslintrc.json:
```diff
@@ -0,0 +1,15 @@
+{
```
All these changes can be moved to root config file, similar to:

https://github.com/WordPress/gutenberg/blob/4d1c083f011a4435c5f12df8d1c14a333df1758c/.eslintrc.js#L433-L438

creativecoder replied:
Nice, done in https://github.com/WordPress/gutenberg/pull/61504/commits/30cab7a38d9e4e9f23c89a6eec4e45a5da348ea3.

gziolo commented on plugins/time-to-read-block/package.json:
```diff
@@ -0,0 +1,30 @@
+{
```
CI reports a few issues with this file:

<img width="1284" alt="Screenshot 2024-05-10 at 08 56 31" src="https://github.com/WordPress/gutenberg/assets/699132/1199968b-a5bf-4a78-856e-7caa487e1fd0">

Some of them might not be applicable because I don't anticipate us publishing these blocks to npm.

gziolo commented on plugins/time-to-read-block/package.json:
```diff
@@ -0,0 +1,30 @@
+		"start": "wp-scripts start"
+	},
+	"devDependencies": {
+		"@wordpress/scripts": "27.7.0"
```
Can we drop it from here? We can use the same script installed in the root of the project. The same applies to some of the script like linting and formatting that are handled for the entire repository already.

gziolo commented on plugins/time-to-read-block/package.json:
```diff
@@ -0,0 +1,30 @@
+		"@wordpress/scripts": "27.7.0"
+	},
+	"dependencies": {
+		"@wordpress/block-editor": "^12.24.0",
```
I think we should reference the local packages that are present in the project, like in other places, example:

https://github.com/WordPress/gutenberg/blob/4d1c083f011a4435c5f12df8d1c14a333df1758c/packages/block-library/package.json#L33-L66

Here, the path will be a bit different:

```suggestion
		"@wordpress/block-editor": "file:../../packages/block-editor",
```

When using explicit version, these quickly will get outdated. By referencing the same packages from the repository, we will always use the latest version. I'm not entirely sure if that is necessary outside of unit tests, but at least we won't have to install these packages from npm.

creativecoder replied:
Yes, that seemed to work as expected. Also done in https://github.com/WordPress/gutenberg/pull/61504/commits/30cab7a38d9e4e9f23c89a6eec4e45a5da348ea3.

creativecoder replied:
@gziolo Thinking through this a bit more, I think we would need to specify the latest version of packages shipped in WordPress, to ensure compatibility.

Block plugins should work without Gutenberg active. Since `@wordpress` packages are transformed to reference `window.wp` as part of the js build process, I believe that production block code could end up using older package versions and unexpectedly break if newer APIs were not available in those older versions.

Does that seem accurate? What do you think?

creativecoder replied:
Then again, it doesn't really seem to matter what package versions are specified here, as either way the actual code in the browser is going to use whatever `window.wp` provides, independent of the packages here in the repo.

The important thing will be running e2e tests without Gutenberg active to ensure the block plugin works with the latest WordPress release even when the Gutenberg plugin is disabled.

gziolo commented on plugins/time-to-read-block/phpcs.xml.dist:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
```
Here, I hope we could also should be able to update the shared config by setting override for the entire folder:

https://github.com/WordPress/gutenberg/blob/trunk/phpcs.xml.dist

Although, I'm not sure how to do it myself. @anton-vlasenko might be able to help.

creativecoder replied:
I looked through the [phpcs configuration docs](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-Ruleset) and can't see a way to apply different rule attributes to different paths in a single command (to use different translation domains for block plugins and the main plugin in the same phpcs check under the `WordPress.WP.I18n` rule). It seems that whenever you redeclare the same ruleset, it overrides any previous instances of the same rule. I believe we could add extra domains to the list, but that doesn't seem like enough, as it wouldn't catch `'gutenberg'` being used in block plugins.

So the best I could come up with so far
- Move most of the rules to a `phpcs.xml.shared` file
- Import the shared rules into the main `phpcs.xml.dist`, and into a separate file for each block plugin
- Use a script when running the `lint:php` command that does a separate run of phpcs for each `phpcs.xml.dist` config

But let me know if I'm missing something or there's a better way to organize this.

gziolo reviewed (commented):
I left a bunch of suggestions on how to move some control to the top of the repository so that whenever we add a new block, we will be updating existing files. I think, the most impactful thing would be getting rid of maintaining the local package lock file since all the packages in most of the cases will be installed in the root `node_modules` folder. By the way, to keep the lock file from the project, you can use the following setting in the `.npmrc` file:

https://github.com/WordPress/gutenberg/blob/4d1c083f011a4435c5f12df8d1c14a333df1758c/packages/block-library/.npmrc#L1

I'm also sure we can figure out a way to be able to build these block from the root of the project, similar to how we run additional build steps for some WP packages using Learn:

https://github.com/WordPress/gutenberg/blob/4d1c083f011a4435c5f12df8d1c14a333df1758c/package.json#L272

We shouldn't reuse Lerna here for a new script, because we don't need to publish anything to npm, but we could start using [npm workspaces](https://docs.npmjs.com/cli/v10/using-npm/workspaces) to achieve the same goal.

creativecoder committed:
Resolve linting issues and integrate phpcs config better with monorepo

creativecoder committed:
Fix merged changes to phpcs.xml.dist

creativecoder committed:
Fix lint:pkg-json errors

creativecoder committed:
Runs plugin commands from root of repo

creativecoder commented:
I've updated this PR so that all plugin commands (like `npm run build`) can be run from the root of the repo.

- For linting/formatting css and js, the root commands already cover the plugin files, so no changes were needed
- For PHP linting/formatting, I added scripts that loop through the plugin directories to run `phpcs` and `phpcbf` triggered by the same `composer run-script` commands we're already using
- For building and plugin-zip generation, the plugin package.json still has separate commands, but they can be run from the root of the repo with npm workspace configuration (e.g. `npm run build --workspace plugins/` to build all plugins). The regular `npm run build` from the repo root has also been updated to trigger this, so that Gutenberg, packages, and plugins can all be built with one command.

creativecoder committed:
Adds e2e test for Time to Read block

creativecoder committed:
Improve e2e tests

- Add custom errors for assertions
- Activate/deactivate plugins in sequence, rather that in parallel, since parallel activation
  seems to cause block not found errors

creativecoder committed:
Adds phpunit tests

- Sets up a phpunit/plugins directory for canonical block plugin tests
- Copies existing time to read block tests into plugins test dir

creativecoder committed:
Adds block plugins to snapshot tests

creativecoder committed:
Uses block supports for textAlign rather than manual attribute

creativecoder committed:
Remove var_dump statement

creativecoder committed:
WIP Adds plugins readme

creativecoder committed:
Removes plugin package lock file

creativecoder committed:
Expands plugins readme with details about file structure, testing, etc

creativecoder committed:
Attempts to get unit tests working in CI

creativecoder committed:
Moves block.json file to make it accessible in phpunit test environment

creativecoder committed:
Updates block.json path glob for fixture tests

creativecoder committed:
Updates readme with block.json location

creativecoder committed:
Adds .npmpackagejsonlintrc.json for plugins directory to use WordPress.org as the author

creativecoder committed:
Excludes plugins/ from core backport check

creativecoder committed:
Updates the description to indicate the reading time is an estimate

creativecoder committed:
Updates backport check exclusion

creativecoder commented:
I've marked this PR ready for review, as I believe all of the technical work needed to try out releasing the Time to Read block in a separate plugin is in place, and CI is all finally passing. (Though the canonical plugins concept likely needs more discussion and a core proposal is still planned.)

cc @gziolo @youknowriad @anton-vlasenko specifically for any more feedback on the monorepo integration, like how npm scripts and tests are set up.

youknowriad commented on composer.json:
```diff
@@ -46,8 +46,8 @@
-		"format": "phpcbf --standard=phpcs.xml.dist --report-summary --report-source",
-		"lint": "phpcs --standard=phpcs.xml.dist",
+		"format": "bash ./scripts/format-php.sh",
+		"lint": "bash ./scripts/lint-php.sh",
```
In general I think we use the `bin` folder for scripts.

creativecoder replied:
Thanks! Updated in https://github.com/WordPress/gutenberg/pull/61504/commits/8e1b30d954f1045c02dab34848b5f6dfbae67af2

youknowriad commented on package.json:
```diff
@@ -268,12 +271,14 @@
 	},
 	"scripts": {
-		"build": "npm run build:packages && wp-scripts build",
+		"build": "npm run build:packages && concurrently 'npm run build:plugins' 'wp-scripts build'",
```
The question to be asked is whether we want to build plugins when we're working on Gutenberg. Building plugins for now is probably very fast but I expect that as we increase the number of canonical blocks, it might become a burden and developers would want to focus on one thing in particular: Gutenberg, a specific block...  An option to build everything, would be handy too but I wonder if it should be the default or not.

creativecoder replied:
My personal expectation is that the primary build command will build the entire project, and that I shouldn't need to run any other commands for a complete build. That said, having a `build:gutenberg` or similar that only builds the main Gutenberg plugin (and packages) seems like a good idea.

youknowriad commented on package.json:
```diff
@@ -268,12 +271,14 @@
-		"build:plugin-zip": "bash ./bin/build-plugin-zip.sh",
+		"build:plugins": "npm run build --workspace ./plugins",
+		"build:plugin-zip": "concurrently 'npm run build:plugin-zip:plugins' 'bash ./bin/build-plugin-zip.sh'",
+		"build:plugin-zip:plugins": "npm run plugin-zip --workspace ./plugins",
```
How does this work, does it build multiple zips for plugins, how to build a single plugin zip?

creativecoder replied:
It loops through the plugins directory and runs the `plugin-zip` script defined in each package json.

The Time to Read Block is using the vanilla `"plugin-zip": "wp-scripts plugin-zip",`, which is what I'd anticipate most plugins doing. The end result is a separate zip for each plugin inside the plugin's directory (e.g. `plugins/time-to-read-block/time-to-read-block.zip`).

youknowriad commented on phpunit/bootstrap.php:
```diff
@@ -64,10 +64,16 @@
+	}
 }
-tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );
+tests_add_filter( 'muplugins_loaded', '_manually_load_plugins' );
```
Do we want to run all the tests of Gutenberg in the same CI job or we want to separate the CI jobs for plugins?

youknowriad replied:
Same for JS and e2e tests.

creativecoder replied:
Good question: it would probably be better to have separate CI jobs, to help ensure all plugins work independently from Gutenberg (and from one another). Though I think all jobs should be run on all PRs, to catch any Gutenberg changes that accidentally break unrelated things in plugins.

youknowriad commented on phpunit/plugins/time-to-read-block/renderBlock.php:
```diff
@@ -0,0 +1,184 @@
+<?php
```
Interesting choice, should it be `phpunit/plugins/time-to-read-block` or `plugins/time-to-read-block/phpunit`?

youknowriad replied:
I guess, similar question for e2e tests.

creativecoder replied:
> Interesting choice, should it be phpunit/plugins/time-to-read-block or plugins/time-to-read-block/phpunit?

That seems reasonable and makes the tests easier to find when they're colocated with the plugins.

That said, I think it's important to be able to run all tests of a particular type using the same command, from the root of the repo, rather than needing to know about separate commands or switch directories to run a complete set of tests. Additionally I think we should avoid duplicating testing configuration as much as possible.

It would also be good to define a separate plugins suite for phpunit, and something similar for other test types, so it's trivial to run plugin tests separately, when desired.

youknowriad replied:
> Additionally I think we should avoid duplicating testing configuration as much as possible.

💯 

In my mind the root level is basically `plugins/gutenberg` but for historic reasons it's in the root. In other words, I like when tests are collocated with the plugin I think, not sure how complex would it be to do that.

 - Leave the configuration in a root folder.
 - Collocate the plugin tests.
 - For Gutenberg it's a special case, maybe keep it root level for now.

youknowriad commented on plugins/README.md:
```diff
@@ -0,0 +1,246 @@
+
+## Block Plugins
+
+Block plugins should follow the [block specifec plugin guidelines](https://developer.wordpress.org/plugins/wordpress-org/block-specific-plugin-guidelines/#block-plugins-and-the-block-directory) and contain only one top-level block per plugin.
```
I can see cases where multiple blocks can be in the same plugin: example a form block with its children blocks. How should these be handled. (Doesn't need to be solved now as the current one doesn't need it though)

creativecoder replied:
I don't think the block directory handles that case, right now. That doesn't have to exclude those plugins from here, though--single block plugins are just the first focus. I'll tweak the language.

youknowriad commented on plugins/README.md:
````diff
@@ -0,0 +1,246 @@
+</ruleset>
+```
+
+#### ESLint
````
Can this override be done in an `.eslintrc.js` within the plugin folder?

creativecoder replied:
I can be, and that's what I had at first, though I changed it on based on @gziolo 's feedback (https://github.com/WordPress/gutenberg/pull/61504#discussion_r1596342389)

youknowriad replied:
Personally I feel collocating is better as most people will just copy/paste a plugin folder and adapt it. Simpler than having to change files elsewhere. 

That said, it's not a big deal either if it stays as is.

youknowriad commented on .wp-env.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"core": "WordPress/WordPress",
-	"plugins": [ "." ],
+	"plugins": [ ".", "./plugins/time-to-read-block" ],
```
If I'm not wrong, this means the plugin will be enabled when running things like performance tests and e2e tests for Gutenberg. While it may be ok for e2e tests, I think we shouldn't enable these plugins in performance tests.

creativecoder replied:
Good catch, I think you're right. I'll try a mapping instead, like we do for the test plugins.

creativecoder replied:
Done in https://github.com/WordPress/gutenberg/pull/61504/commits/417c8dbe58b4996c96f98eb7a843cb106c36ba09

youknowriad reviewed (commented):
Nice work, looks well organized. Left a number of questions.

creativecoder committed:
Moves scripts for php into bin/ folder

creativecoder committed:
Updates block plugin description in plugins readme

creativecoder committed:
Switches to use mappings in wp-env so plugins are not activated by default

creativecoder commented:
Given the [new plan described here](https://github.com/WordPress/gutenberg/issues/58773#issuecomment-2218070785), I'm going to close this PR. But it's in a good state to serve as a reference if and when we return to wanting to use the Gutenberg monorepo for additional plugins.
